### DISCUSSION TITLE
Missing sqlite3_clear_bindings API

### DISCUSSION THREAD

**Discussion by aliasliao** [rhashimoto/wa-sqlite#186] [General]
Accroding to https://www.sqlite.org/c3ref/clear_bindings.html, to clear stmt binding, `sqlite3_clear_bindings` needs to be called.

However, only `reset` is present in [SQLiteAPI](https://rhashimoto.github.io/wa-sqlite/docs/interfaces/SQLiteAPI.html) currently.

P.S. What I want to implement is something like this:
```ts
  async batchRun(sql: string, records: BindingSpec[]): Promise<void> {
    for await (const stmt of this.sqlite3.statements(this.db, sql)) {
      for (const record of records) {
        this.sqlite3.bind_collection(stmt, record);
        await this.sqlite3.step(stmt);
        await this.sqlite3.reset(stmt);
        // need to clear binding here, but there is no such API
      }
    }
  }
```

**Comment by aliasliao** [comment]
Find a workaround:
```ts
  async batchRun(sql: string, records: BindingSpec[]): Promise<void> {
    for await (const stmt of this.sqlite3.statements(this.db, sql)) {
      for (const record of records) {
        this.sqlite3.bind_collection(stmt, record);
        await this.sqlite3.step(stmt);
        await this.sqlite3.reset(stmt);

        // sqlite3_clear_bindings
        const nBindings = this.sqlite3.bind_parameter_count(stmt);
        for (let i = 1; i <= nBindings; ++i) {
          this.sqlite3.bind_null(stmt, i);
        }
      }
    }
  }
```

However, this will significantly impact performance when there are a large number of records. I hope that `sqlite3_clear_bindings` will be added to the API list 🙏.

**Comment by rhashimoto** [comment]
It's a reasonable thing to add. I'll put it on my TODO list, but not at high priority.

The underlying C function is exported by the WebAssembly module (the argument to `SQLite.Factory()` creating the API wrappers), so you should be able to make a call like this as a workaround:

```javascript
  module._sqlite3_clear_bindings(stmt);
```

**Reply by aliasliao** [reply]
Noted, I’ll try it, thanks!

**Reply by aliasliao** [reply]
> It's a reasonable thing to add. I'll put it on my TODO list, but not at high priority.
> 
> The underlying C function is exported by the WebAssembly module (the argument to `SQLite.Factory()` creating the API wrappers), so you should be able to make a call like this as a workaround:
> 
> ```js
>   module._sqlite3_clear_bindings(stmt);
> ```

I tried to implement the `sqlite3.clear_bindings` function following the style of sqlite-api.js, but found it difficult to do so.
This is mainly because some private functions, such as verifyStatement, are not exposed.

 Is it possible to provide some kind of extension mechanism for SQLiteAPI?

For example, implementing `Factory` with `class` and making `verifyStatement`, `verifyDatabase`, and `check` "protected". This way, users can extend the SQLiteAPI like this:
```ts
import { Factory } from 'wa-sqlite/src/sqlite-api';

class MyFactory extends Factory {
    constructor(module) {
        super(module);
    }
	clear_bindings(stmt) {
        this.verifyStatement(stmt);
        ....
    }
}
```

This is also useful to override default implementations of SQLiteAPI. For instance, I would like to override `sqlite3.bind_collection` to enable binding named parameters without requiring a prefix in the object field name:
```ts
  sqlite3.bind_collection(stmt, {
    foo: 42,
    bar: 'hello',
    baz: null,
  });
```

**Reply by sgbeal** [reply]
> For instance, I would like to override sqlite3.bind_collection to enable binding named parameters without requiring a prefix in the object field name:

FWIW, i don't believe that would be possible because the prefix is part of the SQL code and the object keys have to match those exactly. The object being passed to bind_collection() has no way of knowing (without inspecting the SQL, but _That Way Lies Madness_) which prefix is used for any given named parameter.

**Reply by aliasliao** [reply]
> > For instance, I would like to override sqlite3.bind_collection to enable binding named parameters without requiring a prefix in the object field name:
> 
> FWIW, i don't believe that would be possible because the prefix is part of the SQL code and the object keys have to match those exactly. The object being passed to bind_collection() has no way of knowing (without inspecting the SQL, but _That Way Lies Madness_) which prefix is used for any given named parameter.

What I mean is that it's possible to implement this binding at the API level (like [better-sqlite3](https://github.com/WiseLibs/better-sqlite3/blob/master/docs/api.md#binding-parameters)).

By modifying `sqlite3.bind_collection`, this can be achieved:
```diff
sqlite3.bind_collection = function(stmt, bindings) {
  verifyStatement(stmt);
  const isArray = Array.isArray(bindings);
  const nBindings = sqlite3.bind_parameter_count(stmt);
  for (let i = 1; i <= nBindings; ++i) {
    let key = isArray ? i - 1 : sqlite3.bind_parameter_name(stmt, i);
+    
+    if (typeof key === 'string') {
+      key = key.slice(1);
+    }
+
    const value = bindings[key];
    if (value !== undefined) {
      sqlite3.bind(stmt, i, value);
    }
  }
  return SQLite.SQLITE_OK;
};
```

**Reply by sgbeal** [reply]
> sqlite3.bind_parameter_name(stmt, i);

Aha, i had completely forgotten about that function.

**Reply by rhashimoto** [reply]
I'm not planning to support API wrapper extension. You can build/copy your own wrapper support functions, or you can fork.

In the particular case of the internal `verifyStatement()` checks, I don't think it adds much value any more. I thought that it might when the `sqlite3_prepare_v2()` wrapper returned a non-standard object, but now that all statement pointers are provided by the `statements()` iterator in 1.0, I think the mistake of passing a non-statement pointer is much less likely.

**Reply by aliasliao** [reply]
Okay, I tried ignoring `verifyStatement` and directly calling the C API, and it worked. Thanks!
